File: src/http.py
"""http - Promise-based HTTP request helper for opalib.

This module is a Python port of the Roblox/ Lua `http.lua` helper.
It exposes a single `req` function that returns a `Promise` and
provides request timeout handling, late response detection, and
basic response shaping compatible with the Lua implementation.
"""

import threading
import time
import traceback
import warnings
from typing import Any, Callable, Dict, Optional, Union
from urllib.error import HTTPError, URLError
from urllib.request import Request as UrlRequest, urlopen

from src.promise import Promise

TIMEOUT = 120.0
ERR_TIMED_OUT = "Request timed out at %d seconds."
ERR_LATE = "Response arrived late (after %d seconds)."
ERR_HTTP = 'Request received HTTP %d %s "%s"'
ERR_RBX = 'Request received an error: "%s"'


class _LinkNode:
    pass


LL = _LinkNode()
LL.prev = LL
LL.next = LL


def _normalize_arg(args: Dict[str, Any], primary: str, fallback: str) -> Any:
    if primary in args:
        return args[primary]
    return args.get(fallback)


def _make_response(success: bool, status: int, message: str, body: Union[str, bytes]) -> Dict[str, Any]:
    if isinstance(body, bytes):
        try:
            body_text = body.decode("utf-8")
        except Exception:
            body_text = body
    else:
        body_text = body

    return {
        "Success": success,
        "StatusCode": status,
        "StatusMessage": message,
        "Body": body_text,
    }


class Request:
    def __init__(self, args: Dict[str, Any]) -> None:
        self.promise: Optional[Promise] = None
        self.success: Optional[bool] = None
        self.response: Any = None
        self.args = args
        self.next: Optional[Request] = None
        self.prev: Optional[Request] = None
        self.start_t = float("inf")
        self.finish_t = float("inf")
        self.is_active = True
        self.traceback = "".join(traceback.format_stack()[:-1])
        self.msg: Optional[str] = None

    def _send_request(self) -> Dict[str, Any]:
        url = _normalize_arg(self.args, "Url", "url")
        if not url:
            raise ValueError("Missing Url")

        method = _normalize_arg(self.args, "Method", "method") or "GET"
        headers = _normalize_arg(self.args, "Headers", "headers") or {}
        data = _normalize_arg(self.args, "Body", "body")

        request_data: Optional[bytes]
        if data is None:
            request_data = None
        elif isinstance(data, bytes):
            request_data = data
        elif isinstance(data, str):
            request_data = data.encode("utf-8")
        else:
            request_data = str(data).encode("utf-8")

        request = UrlRequest(url, data=request_data, headers=headers, method=method)
        try:
            with urlopen(request, timeout=TIMEOUT) as response:
                body = response.read()
                return _make_response(
                    200 <= response.status < 400,
                    response.status,
                    response.reason or "",
                    body,
                )
        except HTTPError as err:
            body = err.read()
            return _make_response(
                False,
                err.code,
                err.reason or "",
                body,
            )

    def default_throw(self, promise: Promise, arg: Any) -> Any:
        if not self.is_active:
            flight_time = self.finish_t - self.start_t
            if self.success is None:
                warnings.warn(ERR_TIMED_OUT % flight_time)
            else:
                warnings.warn(ERR_LATE % flight_time)

        response = self.response
        if self.success is True and isinstance(response, dict):
            if response.get("Success"):
                warnings.warn(self.msg or "Request completed with an unexpected error.")
            elif response.get("StatusCode") == 401 and response.get("Body") == "Session has expired":
                refresh_callback = self.args.get("refresh")
                if callable(refresh_callback) and promise.predecessor is not None:
                    try:
                        refresh_data = refresh_callback(self.args)
                        if isinstance(refresh_data, dict):
                            self.args.update(refresh_data.get("args", {}))
                            self.args["Headers"] = {
                                **self.args.get("Headers", {}),
                                **refresh_data.get("headers", {}),
                            }
                        promise.Stop()
                        return promise.predecessor.RepeatAsync()
                    except Exception as err:
                        warnings.warn(f"Refresh callback failed: {err}")
                warnings.warn(ERR_HTTP % (response["StatusCode"], response["StatusMessage"], response["Body"]))
            else:
                warnings.warn(ERR_HTTP % (response["StatusCode"], response["StatusMessage"], response["Body"]))
        elif self.success is False:
            warnings.warn(ERR_RBX % response)
        else:
            warnings.warn(self.msg or "Request failed before sending.")

        print(self.traceback)
        return promise.Continue(arg)

    def exec(self, promise: Promise, arg: Any) -> Any:
        self.next = LL
        self.prev = LL.prev
        self.next.prev = self
        self.prev.next = self

        self.start_t = time.time()
        try:
            self.success = True
            self.response = self._send_request()
        except Exception as err:
            self.success = False
            self.response = str(err)
            self.msg = str(err)
        finally:
            self.finish_t = time.time()

        if self.is_active:
            self.prev.next = self.next
            self.next.prev = self.prev
        elif not _normalize_arg(self.args, "AcceptLate", "accept_late"):
            return promise.Throw(self.response, msg=self.msg)

        if self.success is True and isinstance(self.response, dict) and self.response.get("Success"):
            return promise.Continue(self.response)

        return promise.Throw(self.response, msg=self.msg)


def req(args: Dict[str, Any]) -> Promise:
    request = Request(args)
    request.promise = Promise({"request": request})
    request.promise.Silent().Then(request.exec).Else(request.default_throw)
    threading.Timer(0, request.promise.ContinueAsync).start()
    return request.promise


def _timeout_watcher() -> None:
    while True:
        now = time.time()
        current = LL.next
        while current is not LL and current.is_active and now - current.start_t > TIMEOUT:
            current.is_active = False
            current.promise.ThrowAsync()
            current = current.next

        LL.next = current
        current.prev = LL
        time.sleep(0.1)


threading.Thread(target=_timeout_watcher, daemon=True).start()


__all__ = ["req"]

File: http.lua
-- Polaris-Nav, advanced pathfinding as a library and service
-- Copyright (C) 2021 Tyler R. Herman-Hoyer
-- tyler@hoyerz.com
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 3 of the License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program. If not, see <https://www.gnu.org/licenses/>.

local e = _G

local http = game:GetService 'HttpService'

local api = {}

local timeout = 120
local inf = math.huge

local ERR_TIMED_OUT = 'Request timed out at %d seconds.'
local ERR_LATE = 'Response arrived late (after %d seconds).'
local ERR_HTTP = 'Request received HTTP %d %s "%s"'
local ERR_RBX = 'Request received an error: "%s"'

local LL = {}
LL.prev = LL
LL.next = LL

local Request = {}

-- Default error handler. May be called multiple times as errors happen.
function Request:default_throw(args)
	-- Emit if late
	if not self.is_active then
		local flight_time = self.finish_t - self.start_t
		if self.success == nil then
			e.warn(ERR_TIMED_OUT:format(flight_time))
		else
			e.warn(ERR_LATE:format(flight_time))
		end
	end

	-- Emit if there was an error response
	local r = self.response
	if self.success then
		-- lua error after request completed
		if r.Success then
			e.warn(self.msg)
		-- HTTP network error
		elseif r.StatusCode == 401 and r.Body == "Session has expired" then
			local auth = e.store:getState().auth
			return api.refresh {
				token = auth.token;
				id = auth.UserId;
				session = auth.session;
			}:Then(function (p, session)
				e.set_session {
					session = session;
				}
				self.predecessor.args.Headers.session = session
				self:Stop()
				return self.predecessor:RepeatAsync()
			end)
		else
			warn(ERR_HTTP:format(
				r.StatusCode,
				r.StatusMessage,
				r.Body
			))
		end
	-- Roblox network error
	elseif self.success == false then
		e.warn(ERR_RBX:format(r))

	-- lua errors before request was sent
	else
		e.warn(self.msg)
	end

	-- Emit where the request was made
	print(self.traceback)

	-- Continue additional error handlers
	return self:Continue(args, self.msg)
end

-- Sends a request and waits for the response
function Request:exec()
	-- enqueue the request
	self.next = LL
	self.prev = LL.prev
	self.next.prev = self
	self.prev.next = self

	-- send the request
	self.start_t = tick()

	self.success, self.response = pcall(http.RequestAsync, http, self.args)
	self.finish_t = tick()



	if self.is_active then
		self.prev.next = self.next
		self.next.prev = self.prev
	elseif not self.args.accept_late then
		return self:Throw()
	end

	if self.success and self.response.Success then
		return self:Continue()
	else
		return self:Throw()
	end
end

-- Create, enqueue, and send a new request
function api.req(args)
	local req = setmetatable({
		promise = nil;
		success = nil;
		response = nil;
		args = args;
		next = nil;
		prev = nil;
		start_t = inf;
		is_active = true;
		traceback = debug.traceback(nil, 2);
	}, Request)

	-- Promise for handling errors / responses
	req.promise = e.promise(req)
	:Silent()
	:Then(Request.exec)
	:Else(Request.default_throw)

	return req.promise
end

-- Event loop for timing out requests
task.spawn(function()
	while true do
		local t = tick()
		local cur = LL.next
		while cur ~= LL and t - cur.start_t > timeout do
			cur.is_active = false
			cur.promise:ThrowAsync()
			cur = cur.next
		end
		LL.next = cur
		cur.prev = LL

		wait()
	end
end)

return api